### ISSUE TITLE
Failure on migrating from OTRS 5.x 

### ISSUE THREAD

**Issue by sjoswig** [zammad/zammad#826]
### Infos:

* Used Zammad version: zammad-1.3.0-1488535173.5bfe87f.centos7.x86_64
* Operating system: CentOS 7
* Browser + version: Chrome 56

### Expected behavior:

* Migration from OTRS 5.0.16 works as expected

### Actual behavior:

* When starting import I shortly got the message "undefined method `id' for nil:NilClass". 
Tested again from cli and here is the full log:

```
irb(main):004:0> Import::OTRS.start
thread#-: Start import...
Setting.cache: set cache, 750687563
thread#-: POST: http://myhostname/otrs/public.pl?Action=ZammadMigrator
thread#-: PARAMS: {:Action=>"ZammadMigrator", :Key=>"__DELETED__"}
thread#-: loading SysConfig...
thread#-: POST: http://myhostname/otrs/public.pl?Action=ZammadMigrator
thread#-: PARAMS: {:Subaction=>"Export", :Object=>"SysConfig", :Limit=>"", :Offset=>"", :Diff=>0, :Action=>"ZammadMigrator", :Key=>"__DELETED__"}
Setting.reset_cache: set new cache, 102699725
Setting.set(http_type, "http")
Setting.reset_cache: set new cache, 955706322
Setting.set(organization, "NWO")
Setting.reset_cache: set new cache, 602089377
Setting.set(system_id, "01")
Setting.reset_cache: set new cache, 300809925
Setting.set(ticket_hook, "Ticket#")
Setting.reset_cache: set new cache, 542996730
Setting.set(ticket_number, "Ticket::Number::Date")
Setting.reset_cache: set new cache, 911678440
Setting.set(ticket_number_date, {:checksum=>true})
thread#-: loading DynamicField...
Setting.cache: set cache, 911678440
thread#-: POST: http://myhostname/otrs/public.pl?Action=ZammadMigrator
thread#-: PARAMS: {:Subaction=>"Export", :Object=>"DynamicField", :Limit=>"", :Offset=>"", :Diff=>0, :Action=>"ZammadMigrator", :Key=>"__DELETED__"}
thread#-: ERROR: Unsupported dynamic field object type 'Article' for dynamic field 'ArticleFreeText1'
thread#-: loading State...
thread#-: POST: http://myhostname/otrs/public.pl?Action=ZammadMigrator
thread#-: PARAMS: {:Subaction=>"Export", :Object=>"State", :Limit=>"", :Offset=>"", :Diff=>0, :Action=>"ZammadMigrator", :Key=>"__DELETED__"}
thread#-: update Ticket::State.find_by(id: 6)
NoMethodError: undefined method `id' for nil:NilClass
	from /opt/zammad/lib/import/otrs/state.rb:61:in `state_type_id'
	from /opt/zammad/lib/import/otrs/state.rb:54:in `map'
	from /opt/zammad/lib/import/otrs/state.rb:25:in `import'
	from /opt/zammad/lib/import/otrs/state.rb:19:in `initialize'
	from /opt/zammad/lib/import/base_factory.rb:37:in `new'
	from /opt/zammad/lib/import/base_factory.rb:37:in `create_instance'
	from /opt/zammad/lib/import/base_factory.rb:11:in `block in import_action'
	from /opt/zammad/lib/import/base_factory.rb:41:in `each'
	from /opt/zammad/lib/import/base_factory.rb:41:in `import_loop'
	from /opt/zammad/lib/import/otrs/state_factory.rb:22:in `import_loop'
	from /opt/zammad/lib/import/base_factory.rb:9:in `import_action'
	from /opt/zammad/lib/import/transaction_factory.rb:10:in `block in import'
	from /opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/connection_adapters/abstract/database_statements.rb:213:in `block in transaction'
	from /opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/connection_adapters/abstract/transaction.rb:184:in `within_new_transaction'
	from /opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/connection_adapters/abstract/database_statements.rb:213:in `transaction'
	from /opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/transactions.rb:220:in `transaction'
	from /opt/zammad/lib/import/transaction_factory.rb:9:in `import'
	from /opt/zammad/lib/import/otrs.rb:141:in `import_action'
	from /opt/zammad/lib/import/otrs.rb:72:in `import'
	from /opt/zammad/lib/import/otrs.rb:155:in `updateable_objects'
	from /opt/zammad/lib/import/otrs.rb:43:in `start'
	from (irb):4
	from /opt/zammad/vendor/bundle/ruby/2.3.0/gems/railties-4.2.7.1/lib/rails/commands/console.rb:110:in `start'
	from /opt/zammad/vendor/bundle/ruby/2.3.0/gems/railties-4.2.7.1/lib/rails/commands/console.rb:9:in `start'
	from /opt/zammad/vendor/bundle/ruby/2.3.0/gems/railties-4.2.7.1/lib/rails/commands/commands_tasks.rb:68:in `console'
	from /opt/zammad/vendor/bundle/ruby/2.3.0/gems/railties-4.2.7.1/lib/rails/commands/commands_tasks.rb:39:in `run_command!'
	from /opt/zammad/vendor/bundle/ruby/2.3.0/gems/railties-4.2.7.1/lib/rails/commands.rb:17:in `<top (required)>'
	from /opt/zammad/bin/rails:4:in `require'
	from /opt/zammad/bin/rails:4:in `<main>'
```

### Steps to reproduce the behavior:

* Fresh Installation
* Try to import from OTRS 5.0.16

**Comment by thorsteneckel** [comment]
Hi @sjoswig - sorry for the delay. Could you please provide the result of the following URL call with your credentials and domain instead of the dummy data:

`http://your.tld/otrs/public.pl?Action=ZammadMigrator;Key=0123456789;Subaction=Export;Object=State`

That would help allot. You can also send it to us to support@zammad.com. 

Thanks

**Comment by sjoswig** [comment]
```
{"Result":[{"ChangeTime":"2010-01-05 16:53:13","ID":"6","ValidID":"1","TypeID":"4","CreateTime":"2010-01-05 16:53:13","TypeName":"pending reminder","Comment":"Ticket is pending for agent reminder.","Name":"pending reminder"},{"ChangeTime":"2010-10-19 17:42:00","ID":"11","ValidID":"1","TypeID":"8","CreateTime":"2010-10-07 16:00:28","TypeName":"pending customer","Comment":"","Name":"warten auf Kunde"},{"ChangeTime":"2010-01-08 14:40:07","ID":"3","ValidID":"1","TypeID":"3","CreateTime":"2010-01-05 16:53:13","TypeName":"closed","Comment":"Das Ticket wurde ohne Lösung geschlossen","Name":"closed unsuccessful"},{"ChangeTime":"2010-01-05 16:53:13","ID":"7","ValidID":"1","TypeID":"5","CreateTime":"2010-01-05 16:53:13","TypeName":"pending auto","Comment":"Ticket is pending for automatic close.","Name":"pending auto close+"},{"ChangeTime":"2010-01-05 16:53:13","ID":"9","ValidID":"1","TypeID":"7","CreateTime":"2010-01-05 16:53:13","TypeName":"merged","Comment":"State for merged tickets.","Name":"merged"},{"ChangeTime":"2010-11-10 09:28:41","ID":"12","ValidID":"1","TypeID":"9","CreateTime":"2010-11-10 09:28:41","TypeName":"pending external","Comment":"","Name":"warten auf Extern"},{"ChangeTime":"2010-01-08 14:40:13","ID":"2","ValidID":"1","TypeID":"3","CreateTime":"2010-01-05 16:53:13","TypeName":"closed","Comment":"Das Ticket wurde erfolgreich geschlossen","Name":"closed successful"},{"ID":"8","ChangeTime":"2010-01-05 16:53:13","ValidID":"1","TypeID":"5","CreateTime":"2010-01-05 16:53:13","TypeName":"pending auto","Comment":"Ticket is pending for automatic close.","Name":"pending auto close-"},{"ChangeTime":"2010-01-05 16:53:13","ID":"1","ValidID":"1","TypeID":"1","CreateTime":"2010-01-05 16:53:13","TypeName":"new","Comment":"New ticket created by customer.","Name":"new"},{"ChangeTime":"2010-01-05 16:53:13","ID":"4","ValidID":"1","TypeID":"2","CreateTime":"2010-01-05 16:53:13","TypeName":"open","Comment":"Open tickets.","Name":"open"},{"ChangeTime":"2011-10-28 11:31:04","ID":"10","ValidID":"1","TypeID":"3","CreateTime":"2010-08-03 11:16:22","TypeName":"closed","Comment":"ticket is closed with workaround","Name":"closed with workaround"},{"ID":"5","ChangeTime":"2010-01-05 16:53:13","ValidID":"1","TypeID":"6","CreateTime":"2010-01-05 16:53:13","TypeName":"removed","Comment":"Customer removed ticket.","Name":"removed"}],"Notice":"zammad migrator 1.0","Success":1}
```

**Comment by thorsteneckel** [comment]
Thanks for the super fast response! The issue is caused by OTRS state types that are not available in an OTRS default installation. Default types are:

```
new
open
closed
pending reminder
pending auto
removed
merged
```

Your list contains the additional, unknown types:

```
pending customer
pending external
```

These are currently not supported by the OTRS import. 

A way to work around this would be to do the import via the [command line](https://docs.zammad.org/en/latest/migration-otrs.html#import-via-command-line) and running the following commands before starting the import:

```
Ticket::StateType.create_if_not_exists(name: 'pending customer')
Ticket::StateType.create_if_not_exists(name: 'pending external')
```

It may be needed to change the state type of the assigned states to `pending action` and set some of the other attributes like e.g. `ignore_escalation: true` ...

Let me know if that works for you.

**Comment by sjoswig** [comment]
Hi Thorsten,

here is the error output when trying to set command:

```
irb(main):005:0> Ticket::StateType.create_if_not_exists(name: 'pending customer')
Setting.cache: set cache, 712849531
ActiveRecord::StatementInvalid: PG::NotNullViolation: FEHLER:  NULL-Wert in Spalte »updated_by_id« verletzt Not-Null-Constraint
DETAIL:  Fehlgeschlagene Zeile enthält (9, pending customer, null, null, null, 2017-03-22 08:59:46.294986, 2017-03-22 08:59:46.294).
: INSERT INTO "ticket_state_types" ("name", "created_at", "updated_at") VALUES ($1, $2, $3) RETURNING "id"
	from /opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/connection_adapters/postgresql_adapter.rb:602:in `exec_prepared'
	from /opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/connection_adapters/postgresql_adapter.rb:602:in `block in exec_cache'
	from /opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/connection_adapters/abstract_adapter.rb:484:in `block in log'
	from /opt/zammad/vendor/bundle/ruby/2.3.0/gems/activesupport-4.2.7.1/lib/active_support/notifications/instrumenter.rb:20:in `instrument'
	from /opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/connection_adapters/abstract_adapter.rb:478:in `log'
	from /opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/connection_adapters/postgresql_adapter.rb:601:in `exec_cache'
	from /opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/connection_adapters/postgresql_adapter.rb:585:in `execute_and_clear'
	from /opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/connection_adapters/postgresql/database_statements.rb:160:in `exec_query'
	from /opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/connection_adapters/postgresql/database_statements.rb:192:in `exec_insert'
	from /opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/connection_adapters/abstract/database_statements.rb:108:in `insert'
	from /opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/connection_adapters/abstract/query_cache.rb:14:in `insert'
	from /opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/relation.rb:64:in `insert'
	from /opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/persistence.rb:524:in `_create_record'
	from /opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/counter_cache.rb:139:in `_create_record'
	from /opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/locking/optimistic.rb:75:in `_create_record'
	from /opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/attribute_methods/dirty.rb:133:in `_create_record'
... 23 levels...
	from /opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/connection_adapters/abstract/database_statements.rb:213:in `transaction'
	from /opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/transactions.rb:220:in `transaction'
	from /opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/transactions.rb:348:in `with_transaction_returning_status'
	from /opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/transactions.rb:286:in `block in save'
	from /opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/transactions.rb:301:in `rollback_active_record_state!'
	from /opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/transactions.rb:285:in `save'
	from /opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/persistence.rb:34:in `create'
	from /opt/zammad/app/models/application_model/handles_creates_and_updates.rb:69:in `create_if_not_exists'
	from (irb):5
	from /opt/zammad/vendor/bundle/ruby/2.3.0/gems/railties-4.2.7.1/lib/rails/commands/console.rb:110:in `start'
	from /opt/zammad/vendor/bundle/ruby/2.3.0/gems/railties-4.2.7.1/lib/rails/commands/console.rb:9:in `start'
	from /opt/zammad/vendor/bundle/ruby/2.3.0/gems/railties-4.2.7.1/lib/rails/commands/commands_tasks.rb:68:in `console'
	from /opt/zammad/vendor/bundle/ruby/2.3.0/gems/railties-4.2.7.1/lib/rails/commands/commands_tasks.rb:39:in `run_command!'
	from /opt/zammad/vendor/bundle/ruby/2.3.0/gems/railties-4.2.7.1/lib/rails/commands.rb:17:in `<top (required)>'
	from /opt/zammad/bin/rails:4:in `require'
	from /opt/zammad/bin/rails:4:in `<main>'
```

**Comment by thorsteneckel** [comment]
Hi @sjoswig - sorry forgot one command. Here's the complete list:

```ruby
UserInfo.current_user_id = 1
Ticket::StateType.create_if_not_exists(name: 'pending customer')
Ticket::StateType.create_if_not_exists(name: 'pending external')
```

**Comment by sjoswig** [comment]
Hi @thorsteneckel,

now he did a lot of importing stuff I think but it is not possible for me to log into zammad because I always come to import website.

I did a 
```
[root@zammad ~]# zammad run rails c
Loading production environment (Rails 4.2.7.1)
irb(main):001:0> Setting.set('import_mode', false)
Setting.reset_cache: set new cache, 189628635
Setting.set(import_mode, false)
=> true
irb(main):002:0> Setting.set('system_init_done', true)
Setting.reset_cache: set new cache, 183936870
Setting.set(system_init_done, true)
=> true
irb(main):003:0> quit
```

... but this did not help.

Sascha

**Comment by sjoswig** [comment]
Ok @thorsteneckel - now the import seems to be good. Thanks for your help.

**Comment by thorsteneckel** [comment]
Nice! Thanks for your feedback. Closing then.
